### PR TITLE
[moe training] Optimize tensorwise FP8 grouped GEMM

### PR DESCRIPTION
**Pull request by rishisinhanj**
## Summary
This adds a tensorwise FP8 grouped GEMM path for MoE training and optimizes its quantization overhead enough to make the expert backward path materially faster on DeepSeek-V3-sized shapes.

The implementation uses single-scalar tensorwise scaling, caches/materializes the layouts needed by `torch._scaled_grouped_mm`, and replaces the slow layout materialization plus contention-heavy tensorwise amax path with tuned Triton kernels. The dual-layout 2D quantizer now writes row-major and column-major FP8 outputs from tiled blocks, while the amax path uses a staged two-kernel reduction for DeepSeek-sized tensors with an atomic fallback for tensors with too many partial blocks.

## What changed
- Add TensorWise FP8 grouped GEMM plumbing and recipe selection for MoE training.
- Add fused Triton 2D and 3D tensorwise quantization kernels, including row-major and column-major layout materialization for grouped GEMM requirements.
- Add a tiled 2D dual-layout quantization kernel so one pass materializes both grouped-GEMM layouts without a separate copy/transpose path.
- Add a staged tensorwise amax reduction and route both 2D activation/grad quantization and 3D expert-weight quantization through it.
- Add focused correctness tests for staged amax vs the prior atomic path, 2D and dual-layout quantization parity, and grouped-GEMM forward/backward with nonuniform offsets.
- Add a DeepSeek-V3 MoE grouped-GEMM benchmark covering BF16, rowwise FP8, tensorwise FP8, fwd/bwd/fwd+bwd, JSON output, and optional trace capture.

## Performance notes
Tiled dual-layout quantization removes the previous explicit layout materialization overhead. On the original TensorWise quantization benchmark, the tiled kernel reduced the dual-layout quantization path from roughly millisecond-scale PyTorch/layout work to a single fused tiled Triton path for the FP8 row-major and column-major outputs.

Focused staged-amax benchmark on MI355X, DeepSeek bs=2 local shapes:

| Target | Shape | Amax Speedup | 2D Quant Speedup | Dual-Layout Quant Speedup |
|---|---:|---:|---:|---:|
| gate/up input | `65536x7168` | `1.86x` | `1.17x` | `1.19x` |
| gate/up grad out | `65536x2048` | `1.36x` | `1.09x` | `1.06x` |
| down input | `65536x2048` | `1.36x` | `1.09x` | `1.06x` |
| down grad out | `65536x7168` | `1.87x` | `1.17x` | `1.19x` |

DeepSeek-V3 grouped GEMM microbenchmark, batch size 1, sequence length 4096, EP=8:

| GEMM | BF16 fwd+bwd | Rowwise fwd+bwd | Tensorwise fwd+bwd |
|---|---:|---:|---:|
| gate | `2305.46 us` | `2217.52 us` | `1810.45 us` |
| up | `2306.42 us` | `2254.58 us` | `1816.89 us` |
| down | `2128.49 us` | `2101.83 us` | `1742.85 us` |

TorchTitan E2E, DeepSeek-V3 671B 4-layer, 4 GPUs, `activation_checkpoint.mode=none`, averages exclude step 1 warmup and step 8 profiler overhead:

| Config | Avg TPS | Avg TFLOPS |
|---|---:|---:|
| BF16 | `10604` | `250.63` |
| Rowwise FP8 | `9970` | `235.63` |
| Tensorwise FP8 | `10208` | `241.25` |

## Test plan
- `python -m py_compile benchmarks/prototype/moe_training/fp8_tensorwise/bench_fp8_grouped_mm_deepseek_v3.py`
- `PYTHONPATH=/workspace/alex_tensorwise_microbench/ao_tensorwise_fork pytest -q test/prototype/moe_training/test_fp8_tensorwise_grouped_mm.py` (`7 passed` on MI355X)
- Ran DeepSeek-V3 grouped GEMM benchmark with `--batch-size 1 --seq-len 4096 --ep 8 --backward`.
- Ran TorchTitan E2E BF16 vs Rowwise vs Tensorwise with traces saved for step 7.

### CODE REVIEW

File: torchao/prototype/moe_training/config.py [hunk 1]
```diff
@@ -60,6 +61,9 @@ class Float8TrainingOpConfig(TrainingOpBaseConfig):
     # Output dtype for the FP8 grouped GEMMs.
     out_dtype: Optional[torch.dtype] = torch.bfloat16
 
+    # Scaling granularity: AXISWISE (rowwise) or TENSORWISE.
+    scaling_granularity: ScalingGranularity = ScalingGranularity.AXISWISE
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/config.py]: can we just combine this wtih `float8_linear_recipe` so both linear and grouped_mm use the same recipe? is there a reason to keep them separate? enum for one vs string literal for the other is inconsistent as well. i prefer the Literal

Comment (rishisinhanj) [reply torchao/prototype/moe_training/config.py]: @danielvegamyhre done

**Comment by rishisinhanj** [comment]
Hi, @danielvegamyhre, is there any other steps to take for this PR to be merged into upstream?

**Comment by danielvegamyhre** [comment]
@rishisinhanj if i'm reading your e2e training table correctly, it seems like avg tflops/sec utilization actually goes down using this, versus bf16?

given numerical accuracy issues with tensorwise scaling, i'd suggest working on ROCM mxfp8 support for dynamically quantized grouped_mm, rather than adding fp8 tensorwise support

**Comment by rishisinhanj** [comment]
@danielvegamyhre Thanks for the review! yes, we're still not beating bf16 E2E but the GEMM + quantization is improved from before and faster than BF16.  We've narrowed down some other aspects in the E2E such as backward indexing in the permute path that takes much more time in FP8 vs BF16 due to padding constraints, and will soon make a PR that should help with E2E performance for FP8 rowwise and tensorwise. I wanted to separate the PRs for reviewer clarity. 

The goal of landing this is to have a working ROCm tensorwise path, which we can compare to rowwise and new MXFP8 implementations. But we are certainly shifting our focus to MXFP8 for ROCm as you suggested, and not putting too much more focus on tensorwise.
